src/smk/tool/geomark/tool-geomark.js

```diff
@@ -211,14 +211,31 @@ include.module( 'tool-geomark', [
             }
 
             this.toGeomark = function(geomarkInfo, drawingLayer) {
+                const geomarkProperties = this.getGeomarkProperties(geomarkInfo);
                 return {
-                    id: geomarkInfo.id || geomarkInfo.properties.id,
-                    url: geomarkInfo.url || geomarkInfo.properties.url,
+                    id: geomarkInfo.id || geomarkProperties.id || this.showCouldNotLoadMessage(),
```

Comment:
I understand that you want to show an error message when a Geomark can't be created with an id or url. The `showCouldNotLoadMessage` function will return undefined by default, resulting in your Geomark object having an undefined value for id, url or both. Is this your intention? Or should `toGeomark` be returning undefined if id/url is undefined?

Reply:
Sorry, I'm not explaining myself very well. You're right in that it doesn't matter if the `undefined` value being assigned to the id or url is coming from the return value of the function, or `geomarkInfo.id`, 'geomarkInfo.url` etc...

I may be incorrect, but from my point of view, what matters is that the `toGeomark` function is returning an object that would appear to be in an invalid state with undefined `id` and/or `url` properties. The callers of the `toGeomark` function then add this invalid (again, my opinion) geomark to the self.geomark collection.

My suggestion would be something like this:
```
this.toGeomark = function(geomarkInfo, drawingLayer) {
    const geomarkProperties = this.getGeomarkProperties(geomarkInfo);
    const id = geomarkInfo.id || geomarkProperties.id;
    const url = geomarkInfo.url || geomarkProperties.url;

    if (id && url) {
      return {
          id,
          url,
          drawingLayer: drawingLayer,
          isVisible: true
     };

     this.showCouldNotLoadMessage();
     return undefined; 
}
```

Then at lines 271 and 304 where `toGeomark` is called, I would check if the value being returned is undefined and act accordingly.

Alternatively, if you're confident that there will always be id and url values available, simply remove your call to `showCouldNotLoadMessage`.

Reply:
Thanks for your suggestion.

My first thought was to update both callers to check for undefined values and handle undefined values by calling showCouldNotLoadMessage(). But on further thought, I like your suggestion that toGeomark() return either a response without any undefined properties, or "undefined" - either/or. Then calling code can just add an "if" block before proceeding. So I think I'll go with that.